app/uk/gov/hmrc/apiplatformmicroservice/thirdpartyapplication/services/ApplicationIdsForCollaboratorFetcher.scala

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.apiplatformmicroservice.thirdpartyapplication.services
+
+import javax.inject.{Inject, Singleton}
+import uk.gov.hmrc.apiplatformmicroservice.thirdpartyapplication.connectors.{ProductionThirdPartyApplicationConnector, SandboxThirdPartyApplicationConnector}
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ApplicationIdsForCollaboratorFetcher @Inject()(sandboxTpaConnector: SandboxThirdPartyApplicationConnector,
+                                                     productionTpaConnector: ProductionThirdPartyApplicationConnector)
+                                                    (implicit ec: ExecutionContext) {
+
+  def apply(email: String)(implicit hc: HeaderCarrier): Future[Seq[String]] = {
+    for {
+      sandbox <- sandboxTpaConnector.fetchApplicationsByEmail(email)
```

Comment:
These two should happen concurrently and need declaring outside of the for comp.  Otherwise a broken sandbox (failed future) will break any prod stuff.

Reply:
Yeah we should declare them outside so they run concurrently.

The only thing is when we do concatenate the results, are we saying that if one of the futures failed we just return the app IDs from the environment that worked? That's fine, but it will result in cases where the user wouldn't be able to see private APIs that they should see (as they'd be whitelisted in the environment that failed). Would that be a problem?

What about if both environments fail? Should we just return an empty list (so, no whitelisting possible) so the user can at least see the public APIs?

Reply:
I think in other places we've assumed any principal failure is genuine snafu.  We've sort of accepted that failures in subordinate are not good but should not stop the api/page from displaying what it can.